utils/exceptions.cc

```diff
@@ -63,3 +66,18 @@ bool should_stop_on_system_error(const std::system_error& e) {
     }
     return true;
 }
+
+std::atomic<bool> abort_on_internal_error{false};
```

Comment:
does that have to be atomic?
`volatile bool` should work too, no?

Reply:
I think volatile would work here but would be super fragile. If two threads access volatile bool at the same time and at least one is a write then this is undefined behaviour. This won't happen in the current state of the code but would be a very hard to find bug in the future if the startup code gets rearranged. For example, when we start reloading configuration without restarts.

Reply:
BTW - I checked the following program:
#if 0
#include <atomic>
std::atomic<bool> stop_(false);
#else
volatile bool stop_ = false;
#endif

int main(int argc, char **argv)
{
    long n = 1000000000;
    while (!stop_) {
        if (--n < 0)
            stop_ = true;
    }

    return 0;
}
result for atomic:
real	0m0.322s
user	0m0.319s
sys	0m0.003s
result for volatile:
real	0m0.662s
user	0m0.660s
sys	0m0.001s

but it can be wrong because there's only one thread so -O3 may optimize out synchronization.

Reply:
@bhalevy AFAIK `volatile` doesn't provide guarantees for atomicity and multi-threaded access. What's wrong with `std::atomic`?

Reply:
volatile will work here because there's no race on this variable at the moment. We write to it before we start the concurrent reads and concurrent reads are ok because they don't change the value. Reads will see the write done at the beginning because volatile guarantees that they always go to the memory for the value of this variable.

I still think that using volatile instead of atomic here is dangerous in the long run and I'm not sure whether there's any performance benefit of doing it at all.

Reply:
Performance of internal error handling doesn't matter.
If it was, we could use `std::atomic` with relaxed ordering, which adds no overhead.

Reply:
@tgrabiec I just wondered if there's a requirement to use std::atomic that I missed.  I'm ok with using it.

Reply:
Anything that is used cross-thread without extra synchronization should be atomic<>.